src/blacksmith/blacksmith.ts

```diff
@@ -0,0 +1,469 @@
+  },
+);
+
+function playIronClickSound(location: Location) {
+  location.world.playSound(
+    location,
+    Sound.ITEM_TRIDENT_HIT,
+    SoundCategory.PLAYERS,
+    0.7,
+    1,
+  );
+}
+
+async function swingHammer(player: Player, location: Location) {
+  hotIronParticle(location);
+  await wait(0.4, 'seconds');
+  hotIronParticle(location);
+  player.swingMainHand();
+  await wait(0.4, 'seconds');
+  hotIronParticle(location);
+  player.swingMainHand();
+  await wait(0.4, 'seconds');
+}
+
+function hotIronParticle(location: Location) {
+  location.world.spawnParticle(Particle.LAVA, location.add(0, -0.2, 0), 3);
+
+  // Alternative particle effect
+  /*
+  location.world.spawnParticle(
+    Particle.SMOKE_NORMAL,
+    location,
+    0,
+    (Math.random() - 0.5) * 0.2,
+    0,
+    (Math.random() - 0.5) * 0.2,
+  );
+  */
+}
+
+function playFlameParticle(location: Location) {
+  location.world.spawnParticle(Particle.LAVA, location, 3);
+}
+
+function playHammeringSounds(location: Location) {
+  const pitch = Math.random() * 0.2 + 0.8;
+  location.world.playSound(location, Sound.BLOCK_ANVIL_USE, 0.6, pitch);
+}
+
+// Take the molten iron from the anvil with pliers
+Pliers.event(
+  PlayerInteractEntityEvent,
+  (event) => event.player.inventory.itemInMainHand,
+  async (event) => {
+    if (event.hand !== EquipmentSlot.HAND) return;
+    const clicked = event.rightClicked;
+    if (clicked.type !== EntityType.ITEM_FRAME) return;
+    const frame = clicked as ItemFrame;
+    const item = frame.item;
+    if (item.type !== MOLTEN_MATERIAL) return;
+    if (!item.itemMeta.hasCustomModelData()) return;
+    const player = event.player;
+    event.setCancelled(true);
+
+    const pliers = getPliersForItem(item);
+    if (!pliers) return;
+    player.inventory.itemInMainHand = pliers;
+    frame.remove();
+    playIronClickSound(frame.location);
+  },
+);
+
+// Take the molten iron from ground
+Pliers.event(
+  PlayerInteractEvent,
+  (event) => event.player.inventory.itemInMainHand,
+  async (event) => {
+    if (!isRightClick(event.action)) return;
+
+    const raytrace = event.player.rayTraceBlocks(4);
+    const hitPos = raytrace?.hitPosition;
+    if (!hitPos) return;
+    const world = event.player.world;
+    const entities = world.getNearbyEntities(
+      hitPos.toLocation(world),
+      0.5,
+      0.5,
+      0.5,
+    );
+    for (const entity of entities) {
+      if (entity instanceof Item) {
+        const item = entity.itemStack;
+        if (isMoltenMetal(item)) {
+          // Check because this event was fired 3 times. This line may be deleted later
+          if (!Pliers.check(event.player.inventory.itemInMainHand)) return;
+
+          const pliersWithItem = getPliersForItem(item);
+          if (!pliersWithItem) continue;
+
+          entity.itemStack.amount--;
+          await wait(1, 'ticks');
+          event.player.inventory.itemInMainHand = pliersWithItem;
```

Comment:
This could be used to duplicate pliers, if item is moved by player during one tick (e.g. by hacking). Accidental destruction of other items is also possible, though unlikely.